identity/service.go

```diff
@@ -195,43 +221,108 @@ func (s *Service) ProposeVote(v *ProposeVote) (network.Message, onet.ClientError
+	_, err := s.propagateData(sid.SCRoot.Roster, v, propagateTimeout)
 	if err != nil {
 		return nil, onet.NewClientErrorCode(ErrorOnet, cerr.Error())
 	}
-	if len(sid.Votes) >= sid.Latest.Threshold ||
-		len(sid.Votes) == len(sid.Latest.Device) {
+	votesCnt := len(sid.Proposed.Votes)
+	if votesCnt >= sid.Latest.Threshold ||
+		votesCnt == len(sid.Latest.Device) {
 		// If we have enough signatures, make a new data-skipblock and
 		// propagate it
 		log.Lvl3("Having majority or all votes")
 
 		// Making a new data-skipblock
 		log.Lvl3("Sending data-block with", sid.Proposed.Device)
-		reply, cerr := s.skipchain.StoreSkipBlock(sid.Data, nil, sid.Proposed)
+		reply, cerr := s.skipchain.StoreSkipBlock(sid.SCData, nil, sid.Proposed)
 		if cerr != nil {
 			return nil, cerr
 		}
 		_, msg, _ := network.Unmarshal(reply.Latest.Data)
-		log.Lvl3("SB signed is", msg.(*Config).Device)
+		log.Lvl3("SB signed is", msg.(*Data).Device)
 		usb := &UpdateSkipBlock{
 			ID:     v.ID,
 			Latest: reply.Latest,
 		}
-		_, err = s.propagateSkipBlock(sid.Root.Roster, usb, propagateTimeout)
+		_, err = s.propagateSkipBlock(sid.SCRoot.Roster, usb, propagateTimeout)
 		if err != nil {
 			return nil, onet.NewClientErrorCode(ErrorOnet, cerr.Error())
 		}
-		return &ProposeVoteReply{sid.Data}, nil
+		return &ProposeVoteReply{sid.SCData}, nil
 	}
 	return nil, nil
 }
 
+// VerifyBlock makes sure that the new block is legit. This function will be
+// called by the skipchain on all nodes before they sign.
+func (s *Service) VerifyBlock(sbID []byte, sb *skipchain.SkipBlock) bool {
+	// Putting it all in a function for easier error-printing
+	err := func() error {
+		if sb.Index == 0 {
+			log.Lvl4("Always accepting genesis-block")
+			return nil
+		}
+		_, dataInt, err := network.Unmarshal(sb.Data)
+		if err != nil {
+			return errors.New("got unknown packet")
+		}
+		data, ok := dataInt.(*Data)
+		if !ok {
+			return fmt.Errorf("got packet-type %s", reflect.TypeOf(dataInt))
+		}
+		hash, err := data.Hash()
+		if err != nil {
+			return err
+		}
+		// Verify that all signatures work out
+		if len(sb.BackLinkIDs) == 0 {
+			return errors.New("No backlinks stored")
+		}
+		s.identitiesMutex.Lock()
+		defer s.identitiesMutex.Unlock()
+		var latest *skipchain.SkipBlock
+		for _, id := range s.Identities {
+			if id.SCData.Hash.Equal(sb.BackLinkIDs[0]) {
+				latest = id.SCData
+			}
+		}
+		if latest == nil {
+			return errors.New("Backlink was not our latest block")
+		}
+		_, dataInt, err = network.Unmarshal(latest.Data)
+		if err != nil {
+			return err
+		}
+		dataLatest := dataInt.(*Data)
+		sigCnt := 0
+		for dev, sig := range data.Votes {
+			if pub := dataLatest.Device[dev]; pub != nil {
+				if err := crypto.VerifySchnorr(network.Suite, pub.Point, hash, *sig); err != nil {
+					return err
+				}
+				sigCnt++
+			} else {
+				log.Lvl2("Not representative signature detected:", dev)
+			}
+		}
+		if sigCnt >= dataLatest.Threshold || sigCnt == len(dataLatest.Device) {
```

Comment:
For this whole section, it seems possible that if I sucessfully create one valid signature from one valid device, then I can just copy paste it enough time so that the verification pass. Is it possible ? 
If yes, we should add a way to distinct between different signers...